0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+for sequence comprehensions as defined here. The initial encoding will likely be with a function,
+similar to the `Seq#Create` function used for set constructions, that accepts an indicator predicate.
+Axioms will be included to at least state that values are elements of a sequence comprehension iff
+they satisfy this predicate. This property is likely the most important to infer automatically,
+although it says nothing about the ordering or multiplicity of these values.
+
+Since `x | x in C && P(x)` where `C` is a sequence will be a very common pattern, it will likely be worthwhile
+to also include axioms for this pattern to encode the fact that elements in the result
+maintain their ordering from `C`, and that the length of the result and the multiplicities of values
+are all bounded from above by their analogues from `C`.
+
+The translation of `foreach` loops should be reducible to sequence comprehensions
+and other existing features for loops in general,
+as the semantics of such loops can usually be expressed as follows:
+
+```dafny
+// A loop of this form:
+foreach x1, x2, ..., xN | <Range> {
+  <Body>
+}
+
+// Is semantically equivalent to:
+var __s := seq x1, x2, ..., xN | <Range> :: (x1, x2, ..., xN);
+for __i := 0 to |__s| {
+  var (x1, x2, ..., xN) := __s[__i];
+  <Body>
+}
+```
+
+This is not a complete equivalency as the domain of `foreach` loop with explicit `decreaes` clauses
+are allowed to be infinite. It may make sense to support encoding potentially infinite sequences in the Boogie prelude
+even if they are not (yet) supported in Dafny, to close this gap.
+
+## Compilation
+
+The most challenging part of compiling these new features is ensuring the enumeration
+ordering of unordered collection types such as sets and multisets is deterministic when used in a sequence comprehension.
+This is closely related to compiling let-such-that expressions (`var x :| ...; ...`), as such expressions
+are implemented by enumerating through the parameters of a domain to search for a match,
+and hence are influenced by this ordering. The current solution is to require any such expressions
+have only one solution, which is not an option here.
+As illustrated in [this paper](https://easychair.org/publications/paper/dM), the challenge is
+that although the runtime values of `{1, 2}` and `{2, 1}` will compare equal to each other,
+their internal representation may differ and hence their enumeration ordering may as well.
+
+Most if not all current Dafny compilers implement a Dafny `set<T>` with some variation on a hash set,
+which is generally a good default choice since this datatype offers constant-time access on average.
+Although different copies of the same set must place each value into the same hashing bucket, multiple
+values may be assigned to the same bucket, so it is necessary to use a secondary data structure of some kind for each bucket.
+A common simple approach is to use a linked list,
```

Comment:
> the requirement of a deterministic ordering based only on the set of contained elements.

I'm confused, I would think that:
- either you want to enumeration order over a collection to be defined, and then it must be deterministic as well
- or you don't define the enumeration order, and then it doesn't have to be deterministic either since apparently you don't care about it, like because the order doesn't affect the output of your program.

Reply:
You need the order to be deterministic so that, for example, the identity `S == S' ==> (seq x | x in S)[0] == (seq x | x in S')[0]` always holds. That's critical for verification to be sound and agree with compiled runtime behavior. But you don't actually have to know whether `seq x | x in {1, 2}` is actually `[1, 2]` or `[2, 1]`. That's what I mean by "under-specified".

Reply:
Shouldn't the set to seq conversion `(seq x | x in S)` apply an implicit ordering on the seq, which for objects can use the pointer value as the key for that ordering?

Reply:
To me the origin of the problem seems to be that you want to have an ordered iteration (like `(seq x | x in S)`) over an unordered collection.

Why not disallow an ordered iteration over an unordered collection?